File: atividade-03/quick-sort/java/QuickSort.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class QuickSort {

    public static void main(String[] args) {
        String arquivoEntrada = "arq.txt";
        String arquivoSaida = "arq-saida.txt";

        long inicio = System.currentTimeMillis();
        long inicioMemoria = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

        List<Integer> numeros = lerArquivo(arquivoEntrada);
        quickSort(numeros, 0, numeros.size() - 1);
        escreverArquivo(numeros, arquivoSaida);

        long fim = System.currentTimeMillis();
        long fimMemoria = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();

        long tempoExecucao = fim - inicio;
        long memoriaUtilizada = fimMemoria - inicioMemoria;

        System.out.println("Tempo de execução: " + tempoExecucao + " ms");
        System.out.println("Memória utilizada: " + memoriaUtilizada / 1024 + " KB");
    }

    private static List<Integer> lerArquivo(String arquivo) {
        List<Integer> numeros = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(arquivo))) {
            String linha;
            while ((linha = br.readLine()) != null) {
                numeros.add(Integer.parseInt(linha));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return numeros;
    }

    private static void quickSort(List<Integer> numeros, int inicio, int fim) {
        if (inicio < fim) {
            int posicaoPivo = particiona(numeros, inicio, fim);
            quickSort(numeros, inicio, posicaoPivo - 1);
            quickSort(numeros, posicaoPivo + 1, fim);
        }
    }

    private static int particiona(List<Integer> numeros, int inicio, int fim) {
        int pivo = numeros.get(fim);
        int i = inicio - 1;
        for (int j = inicio; j < fim; j++) {
            if (numeros.get(j) <= pivo) {
                i++;
                int temp = numeros.get(i);
                numeros.set(i, numeros.get(j));
                numeros.set(j, temp);
            }
        }
        int temp = numeros.get(i + 1);
        numeros.set(i + 1, numeros.get(fim));
        numeros.set(fim, temp);
        return i + 1;
    }

    private static void escreverArquivo(List<Integer> numeros, String arquivo) {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(arquivo))) {
            for (int numero : numeros) {
                bw.write(numero + "\n");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: atividade-03/quick-sort/python/quick-sort.py
import time
import psutil
import os

def ler_arquivo(arquivo):
    numeros = []
    with open(arquivo, 'r') as file:
        for linha in file:
            numeros.append(int(linha.strip()))
    return numeros

def quick_sort(numeros):
    if len(numeros) <= 1:
        return numeros
    else:
        pivo = numeros[len(numeros) // 2]
        menores = [x for x in numeros if x < pivo]
        iguais = [x for x in numeros if x == pivo]
        maiores = [x for x in numeros if x > pivo]
        return quick_sort(menores) + iguais + quick_sort(maiores)

def escrever_arquivo(numeros, arquivo):
    with open(arquivo, 'w') as file:
        for numero in numeros:
            file.write(f"{numero}\n")

def memoria_utilizada():
    processo = psutil.Process(os.getpid())
    return processo.memory_info().rss

def main():
    arquivo_entrada = 'arq.txt'
    arquivo_saida = 'arq-saida.txt'

    inicio_tempo = time.time()
    inicio_memoria = memoria_utilizada()

    numeros = ler_arquivo(arquivo_entrada)
    numeros_ordenados = quick_sort(numeros)
    escrever_arquivo(numeros_ordenados, arquivo_saida)

    fim_tempo = time.time()
    fim_memoria = memoria_utilizada()

    tempo_execucao = (fim_tempo - inicio_tempo) * 1000  # convertendo para milissegundos
    memoria_utilizada_kb = (fim_memoria - inicio_memoria) / 1024  # convertendo para kilobytes

    print(f"Tempo de execução: {tempo_execucao:.2f} ms")
    print(f"Memória utilizada: {memoria_utilizada_kb:.2f} KB")

main()
